src/sources/postgresql_metrics.rs

```diff
@@ -0,0 +1,851 @@
+            client: None,
+            namespace,
+            tags,
+            datname_filter,
+        };
+        this.build_client().await?;
+        Ok(this)
+    }
+
+    async fn build_client(&mut self) -> Result<(), BuildError> {
+        let client = match &self.tls_config {
+            Some(tls_config) => {
+                let mut builder =
+                    SslConnector::builder(SslMethod::tls_client()).context(TlsFailed)?;
+                builder
+                    .set_ca_file(tls_config.ca_file.clone())
+                    .context(TlsFailed)?;
+                let connector = MakeTlsConnector::new(builder.build());
+
+                let (client, connection) =
+                    self.config
+                        .connect(connector)
+                        .await
+                        .with_context(|| ConnectionFailed {
+                            endpoint: config_to_endpoint(&self.config),
+                        })?;
+                tokio::spawn(connection);
+                client
+            }
+            None => {
+                let (client, connection) =
+                    self.config
+                        .connect(NoTls)
+                        .await
+                        .with_context(|| ConnectionFailed {
+                            endpoint: config_to_endpoint(&self.config),
+                        })?;
+                tokio::spawn(connection);
+                client
+            }
+        };
+
+        let version_row = client
+            .query_one("SELECT version()", &[])
+            .await
+            .with_context(|| VersionFailed {
+                endpoint: config_to_endpoint(&self.config),
+            })?;
+        let version = version_row
+            .try_get::<&str, &str>("version")
+            .with_context(|| VersionFailed {
+                endpoint: config_to_endpoint(&self.config),
+            })?;
+        debug!(message = "Connected to server.", endpoint = %config_to_endpoint(&self.config), server_version = %version);
+
+        self.client = Some(client);
+        Ok(())
+    }
+
+    async fn collect(&mut self) -> stream::BoxStream<'static, Metric> {
+        let build_client = match self.client {
+            Some(_) => Ok(()),
+            None => self.build_client().await,
+        };
+
+        let metrics = match build_client {
+            Ok(()) => self
+                .collect_metrics(self.client.as_ref().expect("should exists at this point"))
+                .await
+                .map_err(|err| err.to_string()),
+            Err(err) => Err(err.to_string()),
+        };
+
+        let (up_value, metrics) = match metrics {
+            Ok(metrics) => (1.0, stream::iter(metrics).boxed()),
+            Err(error) => {
+                self.client = None;
+                emit!(PostgresqlMetricsCollectFailed {
+                    error,
+                    endpoint: self.tags.get("endpoint").expect("should be defined"),
```

Comment:
This too I'd like to see stored somewhere where it is guaranteed to exist rather than pulling from the key/value tags.